lib/patientFinder/utils.rb

```diff
@@ -0,0 +1,165 @@
+            }
+    
+            tests_to_MRNs
+        end
+        
+        #resolves ambiguos candidates to a unique MRN 
+        def self.get_MRN_for_ambiguous(candidate_MRNs)
+            AMBIGUOUS_PATIENTS.each_pair {  |ambiguous_entries, mrn| 
+                return mrn if (candidate_MRNs-ambiguous_entries).empty?
+            }
+            "too ambiguous #{candidate_MRNs}"
+        end
+    
+        #resolves the candidates to the test files (one test file - one candidate - when multiple candidates are found manual anotation is used)
+        def self.get_MRN_for_tests(candidate_MRNs, test_record, test_to_MRN_hash)
+            
+            if(candidate_MRNs.length==1)
+                test_to_MRN_hash[test_record.fileName]=candidate_MRNs.first
+            else
+                test_to_MRN_hash[test_record.fileName]=get_MRN_for_ambiguous(candidate_MRNs)
+            end
+        end    
+    
+        #finds sections with data for a record (out of a array of given sections)    
+        # def self.get_non_empty_sections_for_record(sections, record)
+        #     sections_with_data=Array.new
+        #     sections.each{ |section|
+        #         if !(record[section].nil? || record[section].empty?)
+        #             sections_with_data<< section
+        #         end
+        #     }
+        #     sections_with_data
+        # end
+    
+        #a utility that iterates through a set of master records to find possible matches (candidates) for a test cord
+        def self.get_candidates_for_test(master_records,test_record)
+            
+            candidates=Array.new
+            # candidates
+            master_records.each { |master_record|
+            
+                #simple conditions that will reduce the record pool
+                if(master_record.gender==test_record.gender && master_record.ethnicity[:code]==test_record.ethnicity[:code] && master_record.race[:code] == test_record.race[:code])
+                    
+                    #go inside the sections and collect codes and find if the codes from the test are ALL contained in the the same section of the master 
+                    # sections = [:allergies, :care_goals, :conditions, :encounters, :immunizations, :medical_equipment, :medications, :procedures, :insurance_providers ]
+                
+                    # binding.pry
+                    #TODO - couldn't get this to work on an elegant way - further work needed in order to make it nice
+                    # get_non_empty_sections_for_record(sections, test_record).each{ |sec|
+                    # puts "#{sec}"
+                    # }
+                    # sections_with_data=get_non_empty_sections_for_record(sections, test_record)
+                    
+                    # master_is_match = sections_with_data.all? {|section| Utils::codes_match?(test_record[section],master_record[section])}
```

Comment:
If I recall correctly, the problem is that treating the object like a hash will return nil instead of an empty array if the section doesn't exist. One way you could handle this is to do `test_record.send section` instead of `test_record[section]` and similarly for master_record.